cmd/influx/backup.go

```diff
@@ -1,97 +1,328 @@
+		return err
+	} else if err := f.Sync(); err != nil {
+		return err
+	} else if err := f.Close(); err != nil {
+		return err
+	}
+
+	// Lookup file size.
+	fi, err := os.Stat(path)
 	if err != nil {
 		return err
 	}
+	b.manifest.KV = influxdb.ManifestKVEntry{
+		FileName: b.kvPath(),
+		Size:     fi.Size(),
+	}
 
-	fmt.Printf("Backup ID %d contains %d files\n", id, len(backupFilenames))
+	return nil
+}
 
-	for _, backupFilename := range backupFilenames {
-		dest := filepath.Join(backupFlags.Path, backupFilename)
-		w, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
-		if err != nil {
+func (b *cmdBackupBuilder) backupOrganizations(ctx context.Context) (err error) {
+	// Build a filter if org ID or org name were specified.
+	var filter influxdb.OrganizationFilter
+	if b.org.id != "" {
+		if filter.ID, err = influxdb.IDFromString(b.org.id); err != nil {
 			return err
 		}
-		err = backupService.FetchBackupFile(ctx, id, backupFilename, w)
-		if err != nil {
-			return multierr.Append(fmt.Errorf("error fetching file %s: %v", backupFilename, err), w.Close())
+	} else if b.org.name != "" {
+		filter.Name = &b.org.name
+	}
+
+	// Retrieve a list of all matching organizations.
+	orgs, _, err := b.kvService.FindOrganizations(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	// Back up buckets in each matching organization.
+	for _, org := range orgs {
+		b.logger.Info("Backing up organization", zap.String("id", org.ID.String()), zap.String("name", org.Name))
+		if err := b.backupBuckets(ctx, org); err != nil {
+			return err
 		}
-		if err = w.Close(); err != nil {
+	}
+	return nil
+}
+
+func (b *cmdBackupBuilder) backupBuckets(ctx context.Context, org *influxdb.Organization) (err error) {
+	// Build a filter if bucket ID or bucket name were specified.
+	var filter influxdb.BucketFilter
+	filter.OrganizationID = &org.ID
+	if b.bucketID != "" {
+		if filter.ID, err = influxdb.IDFromString(b.bucketID); err != nil {
+			return err
+		}
+	} else if b.bucketName != "" {
+		filter.Name = &b.bucketName
+	}
+
+	// Retrieve a list of all matching organizations.
+	buckets, _, err := b.kvService.FindBuckets(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	// Back up shards in each matching bucket.
+	for _, bkt := range buckets {
+		if err := b.backupBucket(ctx, org, bkt); err != nil {
 			return err
 		}
 	}
+	return nil
+}
+
+func (b *cmdBackupBuilder) backupBucket(ctx context.Context, org *influxdb.Organization, bkt *influxdb.Bucket) (err error) {
+	b.logger.Info("Backing up bucket", zap.String("id", bkt.ID.String()), zap.String("name", bkt.Name))
+
+	// Lookup matching database from the meta store.
+	dbi := b.metaClient.Database(bkt.ID.String())
+	if dbi == nil {
+		return fmt.Errorf("bucket database not found: %s", bkt.ID.String())
+	}
+
+	// Iterate over and backup each shard.
+	for _, rpi := range dbi.RetentionPolicies {
+		for _, sg := range rpi.ShardGroups {
+			for _, sh := range sg.Shards {
```

Comment:
We need to check if `sg.Deleted()` is `true`; otherwise, the later call to 

```go
b.backupService.BackupShard(ctx, gw, shardID, time.Time{})
```

fails with an internal server error, as the shard does not exist.

I tried locally and was able to successfully backup my bucket:

```
❯ bin/darwin/influx backup -b "my-bucket" ~/tmp/backup
2020-10-30T22:55:36.604939Z	info	Backing up KV store	{"log_id": "0QAwQOc0000", "path": "20201030T225536Z.bolt"}
2020-10-30T22:55:36.609293Z	info	Resources opened	{"log_id": "0QAwQOc0000", "path": "/Users/stuartcarnie/tmp/backup/20201030T225536Z.bolt"}
2020-10-30T22:55:36.609606Z	info	Backing up organization	{"log_id": "0QAwQOc0000", "id": "93b0c30ea94cb097", "name": "my-org"}
2020-10-30T22:55:36.609685Z	info	Backing up bucket	{"log_id": "0QAwQOc0000", "id": "7425a44ac4116e88", "name": "my-bucket"}
2020-10-30T22:55:36.609697Z	info	Backing up shard	{"log_id": "0QAwQOc0000", "id": 1, "path": "20201030T225536Z.s1.tar.gz"}
Error: An internal error has occurred.
See 'influx backup -h' for help
```

Rebuilt with check:

```
❯ go build -mod=readonly -o bin/darwin/influx ./cmd/influx
❯ bin/darwin/influx backup -b "my-bucket" ~/tmp/backup
2020-10-30T23:03:51.072926Z	info	Backing up KV store	{"log_id": "0QAwt_80000", "path": "20201030T230351Z.bolt"}
2020-10-30T23:03:51.176761Z	info	Resources opened	{"log_id": "0QAwt_80000", "path": "/Users/stuartcarnie/tmp/backup/20201030T230351Z.bolt"}
2020-10-30T23:03:51.177363Z	info	Backing up organization	{"log_id": "0QAwt_80000", "id": "93b0c30ea94cb097", "name": "my-org"}
2020-10-30T23:03:51.177517Z	info	Backing up bucket	{"log_id": "0QAwt_80000", "id": "7425a44ac4116e88", "name": "my-bucket"}
2020-10-30T23:03:51.177538Z	info	Backing up shard	{"log_id": "0QAwt_80000", "id": 27, "path": "20201030T230351Z.s27.tar.gz"}
2020-10-30T23:03:51.288670Z	info	Backing up shard	{"log_id": "0QAwt_80000", "id": 77, "path": "20201030T230351Z.s77.tar.gz"}
2020-10-30T23:03:51.318524Z	info	Writing manifest	{"log_id": "0QAwt_80000", "path": "20201030T230351Z.manifest"}
2020-10-30T23:03:51.319545Z	info	Backup complete	{"log_id": "0QAwt_80000"}
```